integration/experiment/ffnet/ffnet.py

```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+#
+#  Copyright (c) 2015 - 2024, Intel Corporation
+#  SPDX-License-Identifier: BSD-3-Clause
+#
+
+'''
+Helper functions for running CPU activity agent experiments.
+'''
+
+import argparse
+import os
+
+import geopmpy.agent
+
+from integration.experiment import launch_util
+from integration.experiment import common_args
+from integration.experiment import machine
+
+def setup_run_args(parser):
+    common_args.setup_run_args(parser)
+    parser.add_argument('--perf-energy-bias', dest='perf_energy_bias',
+                        action='store', default=0,
+                        help='Perf-Energy Bias [0-1] where 0 is perf-sensitive and 1 is energy-efficient')
+    parser.add_argument('--cpu-nn-path', dest='cpu_nn_path',
+                        action='store', default=None,
+                        help='Full path for the CPU NN')
+    parser.add_argument('--gpu-nn-path', dest='gpu_nn_path',
+                        action='store', default=None,
+                        help='Full path for the GPU NN')
+    parser.add_argument('--cpu-fmap-path', dest='cpu_freq_rec_path',
+                        action='store', default=None,
+                        help='Full path for the CPU Frequency Recommender Map')
+    parser.add_argument('--gpu-fmap-path', dest='gpu_freq_rec_path',
+                        action='store', default=None,
+                        help='Full path for the GPU Frequency Recommender Map')
+
+def report_signals():
+    return []
+
+def trace_signals():
+    return []
+
+def setup_env_paths(cpu_nn_path=None, cpu_fmap_path=None, gpu_nn_path=None, gpu_fmap_path=None):
+    if cpu_nn_path is None and gpu_nn_path is None:
+        raise RuntimeError('Must specify cpu-nn-path and/or gpu-nn-path when running ffnet experiment')
+
+    if cpu_nn_path is not None:
+        if os.path.exists(cpu_nn_path):
+            os.environ['GEOPM_CPU_NN_PATH'] = cpu_nn_path
+        else:
+            raise KeyError(f'File cpu-nn-path={cpu_nn_path} does not exist.')
+        if cpu_fmap_path is not None:
+            os.environ['GEOPM_CPU_FMAP_PATH'] = cpu_fmap_path
+        else:
+            raise RuntimeError('Must specify cpu-fmap-path when cpu-nn-path is specified for ffnet experiment')
+
+    if gpu_nn_path is not None:
+        if os.path.exists(gpu_nn_path):
+            os.environ['GEOPM_GPU_NN_PATH'] = gpu_nn_path
+        else:
+            raise KeyError(f'File gpu-nn-path={gpu_nn_path} does not exist.')
+        if gpu_fmap_path is not None:
+            os.environ['GEOPM_GPU_FMAP_PATH'] = gpu_fmap_path
+        else:
+            raise RuntimeError('Must specify gpu-fmap-path when gpu-nn-path is specified for ffnet experiment')
+
+    return
+
+def launch_configs(output_dir, app_conf, phi=0):
+    mach = machine.init_output_dir(output_dir)
+
+    if phi > 1 or phi < 0:
+        raise KeyError('perf-energy-bias must be between 0 and 1 for ffnet experiment.')
```

Comment:
I wasn't sure what error to use here. It's used with the size of each array of frequencies in the freq region map to determine which of the provided frequencies to use. E.g. phi=0.3 and freqs=[0, 1, 2, 3, 4]. freqs has a length of 5, so the index is floor(5*0.3) = 1. If phi > 0 or < 1, the index will be outside the array, so that's why I threw a KeyError. I'm open to something more appropriate, though!